astropy/io/fits/tests/test_connect.py

```diff
@@ -414,6 +414,42 @@ def test_mask_str_on_read(self, tmp_path):
         tab = Table.read(filename, mask_invalid=False)
         assert tab.mask is None
 
+    def test_write_object_tables_with_unified(self, tmp_path):
+        """
+        Write objects with the unified I/O interface.
+        See https://github.com/astropy/astropy/issues/1906
+        """
+        filename = tmp_path / "test_table_object.fits"
+        # Make a FITS table with an object column
+        tab = Table([{"col1": None}])
+        with pytest.raises(TypeError) as err:
+            tab.write(filename)
+        assert err.value.args[0] == (
+            "Column contains not supported object types or "
+            "different types are present in the same column."
+        )
+
+    def test_write_VLA_tables_with_unified(self, tmp_path):
+        """
+        Write VLA objects with the unified I/O interface.
+        See https://github.com/astropy/astropy/issues/11323
+        """
+
+        filename = tmp_path / "test_table_VLA.fits"
+        # Make a FITS table with a variable-length array column
+        a = np.array([45, 30])
+        b = np.array([11, 12, 13])
+        c = np.array([45, 55, 65, 75])
+        var = np.array([a, b, c], dtype=object)
+
+        tabw = Table([{"col1": var}])
+        tabw.write(filename, overwrite=True)
```

Comment:
@saimn I have skipped the writing to a temporary file but now the test fails.
Note that if I test
```py
import numpy as np
from astropy.io import fits
from astropy.table import Table

# Make a FITS table with a variable-length array column
a = np.array([45, 30])
b = np.array([11, 12, 13])
c = np.array([45, 55, 65, 75])
var = np.array([a, b, c], dtype=object)
c1 = fits.Column(name='var', format='PJ()', array=var)
hdu = fits.BinTableHDU.from_columns([c1])
tab = Table.read(hdu)
tab.write('temp.fits', overwrite=True)
```
this is working, I guess because the format is specified when defining the column.

Do you mind clarifying to me, how far we want to support writing an "object" type?
If the test is good right now, I'm not sure I will be able to finalise this PR.

Reply:
@cmarmo - it's because you have an unwanted `[...]` in `Table([{"col1": var}])`:
```
In [10]: Table([{"col1": var}])
Out[10]: 
<Table length=1>
          col1          
       object[3]        
------------------------
[45 30] .. [45 55 65 75]
```
The correct syntax is:
```
In [11]: Table({"col1": var})
Out[11]: 
<Table length=3>
     col1    
    object   
-------------
      [45 30]
   [11 12 13]
[45 55 65 75]
```

Reply:
So what we want to support is converting a `Table` like `t = Table({'a': [[1, 2], [3, 4, 5]]})` (or similar with Numpy arrays) to FITS, by guessing the type of the object array items.

If you use the `io.fits` directly, e.g. `fits.Column(name='var', format='PJ()', array=var)`, you specify the type with `format` and the array is cast to this type.

The difference with the unified io interface (the `table_to_hdu` function) is that it converts the `Table` to a Numpy record array, from which the `BinTableHDU` is created. So basically this:
```
In [2]: t = Table({'a': [[1, 2], [3, 4, 5]]})

In [3]: t
Out[3]: 
<Table length=2>
    a    
  object 
---------
   [1, 2]
[3, 4, 5]

In [4]: tmp = t.as_array()

In [5]: tmp
Out[5]: array([(list([1, 2]),), (list([3, 4, 5]),)], dtype=[('a', 'O')])

In [6]: fits.BinTableHDU(tmp)
Out[6]: <astropy.io.fits.hdu.table.BinTableHDU at 0x7f6ee554e990>
```
And `fits.BinTableHDU(tmp)` uses the code we are modifying here to find/guess the `fits.Column` formats from the recarray dtype.